prog/dftb+/lib_dftb/rangeseparated.F90

```diff
@@ -703,6 +719,180 @@ end subroutine updateHamiltonianBlock
+    call evaluateHamiltonian(this, Smat, Dmat, LRgammaAO, Hlr)
+    HH(:,:) = HH + Hlr
+    this%lrenergy = this%lrenergy + 0.5_dp * sum(Dmat*Hlr)
+
+  contains
+
+    subroutine allocateAndInit(this, iSquare, orb, overlap, densSqr, HH, Smat, Dmat, LRgammaAO)
+
+      !> class instance
+      type(RangeSepFunc), intent(inout) :: this
+
+      !> Position of each atom in the rows/columns of the square matrices. Shape: (nAtom)
+      integer, dimension(:), intent(in) :: iSquare
+
+      !> Orbital information.
+      type(TOrbitals), intent(in) :: orb
+
+      !> Square (unpacked) overlap matrix.
+      real(dp), intent(in) :: overlap(:,:)
+
+      !> Square (unpacked) density matrix
+      real(dp), intent(in) :: densSqr(:,:)
+
+      !> Square (unpacked) Hamiltonian to be updated.
+      real(dp), intent(inout) :: HH(:,:)
+
+      !> Symmetrized square overlap matrix
+      real(dp), intent(out) :: Smat(:,:)
+
+      !> Symmetrized square density matrix
+      real(dp), intent(out) :: Dmat(:,:)
+
+      !> Symmetrized long-range gamma matrix
+      real(dp), intent(out) :: LRgammaAO(:,:)
+
+      integer, allocatable :: get_iat(:)
+
+      integer :: nAtom, nOrb
+      integer :: iAt, G1, G2, mu, nu
+
+      nAtom = size(orb%nOrbAtom,dim=1)
+      nOrb = size(overlap,dim=1)
+
+      allocate(get_iat(nOrb))
+
+      ! Symmetrize Hamiltonian, overlap, density matrices
+      call symmetrizeSquareMatrix(HH)
+      Smat(:,:) = overlap
+      call symmetrizeSquareMatrix(Smat)
+      Dmat(:,:) = densSqr
+      call symmetrizeSquareMatrix(Dmat)
+
+      do mu = 1, nOrb
+        do iAt = 1, nAtom
+          if (mu > iSquare(iAt)-1 .and. mu <= iSquare(iAt+1)-1) then
+            get_iat(mu) = iAt
+          end if
+        end do
+      end do
+
+      ! Get long-range gamma variable
+      LRgammaAO(:,:) = 0.0_dp
+      do mu = 1, nOrb
+        do nu = mu, nOrb
+          ! Find proper atom index
+          G1 = get_iat(mu)
+          G2 = get_iat(nu)
+          LRgammaAO(nu,mu) = this%lrGammaEval(G2,G1)
+          if (nu /= mu) LRgammaAO(mu,nu) = LRgammaAO(nu,mu)
+        end do
+      end do
+
+    end subroutine allocateAndInit
+
+    subroutine evaluateHamiltonian(this, Smat, Dmat, LRgammaAO, Hlr)
+
+      !> class instance
+      type(RangeSepFunc), intent(inout) :: this
+
+      !> Symmetrized square overlap matrix
+      real(dp), intent(in) :: Smat(:,:)
+
+      !> Symmetrized square density matrix
+      real(dp), intent(in) :: Dmat(:,:)
+
+      !> Symmetrized long-range gamma matrix
+      real(dp), intent(in) :: LRgammaAO(:,:)
+
+      !> Symmetrized long-range Hamiltonian matrix
+      real(dp), intent(out) :: Hlr(:,:)
+
+      real(dp), allocatable :: Hmat(:,:)
+      real(dp), allocatable :: tmpMat(:,:)
+
+      integer :: nOrb
+
+      nOrb = size(Smat,dim=1)
+
+      allocate(Hmat(nOrb,nOrb))
+      allocate(tmpMat(nOrb,nOrb))
+
+      Hlr(:,:) = 0.0_dp
+
```

Comment:
You can get the gemm operation to make the copying and assembly of Hlr which simplifies the code a little (but doesn't save much in run time or reduction of variables):

      call gemm(tmpMat, Smat, Dmat)
      call gemm(Hlr, tmpMat, Smat)
      Hlr(:,:) = Hlr * LRGammaAO

      tmpMat(:,:) = LRgammaAO * tmpMat
      call gemm(Hlr, tmpMat, Smat, alpha = 1.0_dp, beta = 1.0_dp)

      Hmat(:,:) = Dmat * LRgammaAO
      call gemm(tmpMat, Smat, Hmat)
      call gemm(Hlr, tmpMat, Smat, alpha = 1.0_dp, beta = 1.0_dp)

      call gemm(tmpMat, Dmat, Smat)
      tmpMat(:,:) = tmpMat * LRgammaAO
      call gemm(Hlr, Smat, tmpMat, alpha = 1.0_dp, beta = 1.0_dp)